### PR TITLE
Feat/pubkeyconverter_refactor

### PR DESCRIPTION
**Pull request by schimih**
## Reasoning behind the pull request
- 
- 
- 
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: outport/process/alteredaccounts/alteredAccountsProvider.go [hunk 1]
```diff
@@ -230,20 +235,25 @@ func (aap *alteredAccountsProvider) addTokensDataForMarkedAccount(
 	return nil
 }
 
-func (aap *alteredAccountsProvider) convertMetaData(metaData *esdt.MetaData) *outportcore.TokenMetaData {
+func (aap *alteredAccountsProvider) convertMetaData(metaData *esdt.MetaData) (*outportcore.TokenMetaData, error) {
 	if metaData == nil {
-		return nil
+		return nil, nil
```
Comment (iulianpascalau) [inline outport/process/alteredaccounts/alteredAccountsProvider.go:240]: this is extremely dangerous. Let's refactor. 
Reason: you test for nil error and then you want to use the object (that might also be nil)

File: state/accountsDB.go [hunk 2]
```diff
@@ -1344,7 +1344,8 @@ func (adb *AccountsDB) snapshotUserAccountDataTrie(
 			ErrChan:    iteratorChannels.ErrChan,
 		}
 		if isSnapshot {
-			address := adb.addressConverter.Encode(account.Address)
+			// TODO: TakeSnapshot should receive address as byte array and encode them before printing
```
Comment (iulianpascalau) [inline state/accountsDB.go:1347]: TODO relevant? Remove or implement.

File: state/accountsDB.go [hunk 3]
```diff
@@ -1453,8 +1454,12 @@ func (adb *AccountsDB) GetStatsForRootHash(rootHash []byte) (common.TriesStatist
 			continue
 		}
 
-		address := adb.addressConverter.Encode(account.Address)
-		collectStats(tr, stats, account.RootHash, address)
+		accountAddress, err := adb.addressConverter.Encode(account.Address)
+		if err != nil {
+			return nil, err
+		}
+		//TODO: collectStats should receive accountAddress as byte array
```
Comment (iulianpascalau) [inline state/accountsDB.go:1461]: TODO relevant?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: cmd/keygenerator/converter/pidPubkeyConverter.go [hunk 4]
```diff
@@ -33,7 +33,7 @@ func (converter *pidPubkeyConverter) Decode(_ string) ([]byte, error) {
 }
 
 // Encode encodes a byte array in its string representation
-func (converter *pidPubkeyConverter) Encode(pkBytes []byte) string {
+func (converter *pidPubkeyConverter) Encode(pkBytes []byte) (string, error) {
 	pidString, err := converter.encode(pkBytes)
 	if err != nil {
 		log.Warn("pidPubkeyConverter.Encode encode",
```
[the remark this answers is not in the document]
Comment (schimih) [reply cmd/keygenerator/converter/pidPubkeyConverter.go:39]: true. deleted. :+1:

File: go.mod [hunk 5]
```diff
@@ -13,13 +13,13 @@ require (
 	github.com/google/gops v0.3.18
 	github.com/gorilla/websocket v1.5.0
 	github.com/mitchellh/mapstructure v1.5.0
-	github.com/multiversx/mx-chain-core-go v1.1.35
+	github.com/multiversx/mx-chain-core-go v1.1.36-0.20230308081722-5262fb09cb9a
```
[the remark this answers is not in the document]
Comment (schimih) [reply go.mod:16]: immediately after review and system test :+1:

File: config/tomlConfig_test.go [hunk 6]
```diff
@@ -493,208 +493,143 @@ func TestEnableEpochConfig(t *testing.T) {
 [EnableEpochs]
     # SCDeployEnableEpoch represents the epoch when the deployment of smart contracts will be enabled
     SCDeployEnableEpoch = 1
-
```
[the remark this answers is not in the document]
Comment (schimih) [reply config/tomlConfig_test.go:496]: changed

File: outport/process/alteredaccounts/alteredAccountsProvider.go (hunk 1 above)
Comment (schimih) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:240]: refactored, my bad

File: node/external/transactionAPI/apiTransactionProcessor.go [hunk 7]
```diff
@@ -668,7 +655,7 @@ func (atp *apiTransactionProcessor) getTxBytesFromStorageByEpoch(hash []byte, ep
 	return nil, transaction.TxTypeInvalid, false
 }
 
-func (atp *apiTransactionProcessor) castObjToTransaction(txObj interface{}, txType transaction.TxType) (*transaction.ApiTransactionResult, error) {
+func (atp *apiTransactionProcessor) castObjToTransaction(txObj interface{}, txType transaction.TxType) *transaction.ApiTransactionResult {
```
[the remark this answers is not in the document]
Comment (schimih) [reply node/external/transactionAPI/apiTransactionProcessor.go:658]: yes.

File: node/external/transactionAPI/apiTransactionProcessor.go [hunk 8]
```diff
@@ -333,10 +323,10 @@ func (atp *apiTransactionProcessor) extractRequestedTxInfo(wrappedTx *txcache.Wr
 		tx.TxFields[nonceField] = wrappedTx.Tx.GetNonce()
 	}
 	if requestedFieldsHandler.HasSender {
```
[the remark this answers is not in the document]
Comment (schimih) [reply node/external/transactionAPI/apiTransactionProcessor.go:325]: true, changed

File: node/external/transactionAPI/apiTransactionProcessor.go [hunk 9]
```diff
@@ -100,6 +100,11 @@ func (atp *apiTransactionProcessor) GetTransaction(txHash string, withResults bo
 
 	tx.Hash = txHash
 	atp.PopulateComputedFields(tx)
+	atp.gasUsedAndFeeProcessor.computeAndAttachGasUsedAndFee(tx)
```
[the remark this answers is not in the document]
Comment (schimih) [reply node/external/transactionAPI/apiTransactionProcessor.go:103]: fixed

File: integrationTests/consensus/consensus_test.go [hunk 10]
```diff
@@ -98,7 +98,7 @@ func initNodesAndTest(
 		}
 	}
 
-	return nodes
+	return nodes, nil
```
[the remark this answers is not in the document]
Comment (schimih) [reply integrationTests/consensus/consensus_test.go:101]: reverted

File: state/accountsDB.go (hunk 2 above)
Comment (schimih) [reply state/accountsDB.go:1347]: implemented already :+1:

File: state/accountsDB.go (hunk 3 above)
Comment (schimih) [reply state/accountsDB.go:1461]: not relevant anymore

File: state/syncer/userAccountsSyncer.go [hunk 11]
```diff
@@ -288,9 +288,9 @@ func (u *userAccountsSyncer) printDataTrieStatistics() {
 	})
 
 	for _, trieStat := range u.largeTries {
-		address := u.pubkeyCoverter.Encode(trieStat.address)
+		trieStatAddress := u.pubkeyCoverter.SilentEncode(trieStat.address, log)
```
[the remark this answers is not in the document]
Comment (schimih) [reply state/syncer/userAccountsSyncer.go:291]: changed

File: cmd/keygenerator/main.go [hunk 12]
```diff
@@ -41,6 +41,7 @@ const minedWalletPrefixKeys = "mined-wallet"
 const nopattern = "nopattern"
 const desiredpattern = "[0-f]+"
 const noshard = -1
+const pubkeyHrp = "erd"
```
[the remark this answers is not in the document]
Comment (schimih) [reply cmd/keygenerator/main.go:44]: noted

File: process/smartContract/process.go [hunk 13]
```diff
@@ -1800,11 +1800,15 @@ func (sc *scProcessor) printScDeployed(vmOutput *vmcommon.VMOutput, tx data.Tran
 			continue
 		}
 
-		scGenerated = append(scGenerated, sc.pubkeyConv.Encode(addr))
+		scAddress := sc.pubkeyConv.SilentEncode(addr, log)
+
```
[the remark this answers is not in the document]
Comment (schimih) [reply process/smartContract/process.go:1804]: deleted

File: process/smartContract/process.go [hunk 14]
```diff
@@ -1800,11 +1800,15 @@ func (sc *scProcessor) printScDeployed(vmOutput *vmcommon.VMOutput, tx data.Tran
...
+		scGenerated = append(scGenerated, scAddress)
 	}
 
+	encodedSndAddr := sc.pubkeyConv.SilentEncode(tx.GetSndAddr(), log)
+
 	log.Debug("SmartContract deployed",
-		"owner", sc.pubkeyConv.Encode(tx.GetSndAddr()),
+		"owner", encodedSndAddr,
```
[the remark this answers is not in the document]
Comment (schimih) [reply process/smartContract/process.go:1811]: changed

File: cmd/node/config/config.toml [hunk 15]
```diff
@@ -599,6 +599,7 @@
     Length = 32
     Type = "bech32"
     SignatureLength = 64
+    Hrp = "erd"
```
[the remark this answers is not in the document]
Comment (schimih) [reply cmd/node/config/config.toml:602]: deleted

File: node/external/transactionAPI/apiTransactionProcessor.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply node/external/transactionAPI/apiTransactionProcessor.go:658]: great

File: cmd/keygenerator/main.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/keygenerator/main.go:44]: ...or a config defined in config.toml file

File: node/node_test.go [hunk 16]
```diff
@@ -875,7 +876,7 @@ func TestNode_GetAllESDTTokensShouldReturnEsdtAndFormattedNft(t *testing.T) {
 	tokens, _, err := n.GetAllESDTTokens(testscommon.TestAddressAlice, api.AccountQueryOptions{}, context.Background())
 	assert.Nil(t, err)
 	assert.Equal(t, 2, len(tokens))
-	assert.Equal(t, esdtData, tokens[esdtToken])
+	assert.Equal(t, esdtData, tokens["TKKR-7q8w9e-"])
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply node/node_test.go:879]: reverted in the next PR :D

**Review by gabi-vuls — APPROVED** [review]
System test passed

**Review by iulianpascalau — APPROVED** [review]
go.mod looks ok :ok_hand:

**Review by andreibancioiu — APPROVED** [review]

**Pull request merged**